cluster/saltbase/salt/node-upgrade-gcp/node-upgrade-gcp.sh

```diff
@@ -0,0 +1,157 @@
+  echo "$(date): ${@}"
+}
+
+# get-metadata curls all keys in WATCH_KEYS from the metadata server, retrying
+# until success, and saves them in CACHE_DIR named by the key and the provided
+# extension.
+#
+# Assumed vars:
+#   WATCH_KEYS
+#   CACHE_DIR
+#
+# Args:
+#  $1 -- extension to use for files (e.g. CACHE_EXT or LATEST_EXT)
+function get-metadata() {
+  local ext="${1}"
+  log "Attempting to get metadata."
+  for k in "${WATCH_KEYS[@]}"; do
+    log "Attempting to get metadata with key '${k}'."
+    local cache_file="${CACHE_DIR}/${k}.${ext}"
+    # We use --retry for some exponential backoff, but this curls until success.
+    until curl -f -s --retry 5 -o "${cache_file}" -H "Metadata-Flavor: Google" \
+      "http://metadata/computeMetadata/v1/instance/attributes/${k}"; do :; done
+    log "Successfully saved metadata with key '${k}' in '${cache_file}'."
+  done
+}
+
+# monitor-upgradability runs until any metadata on the server changes, or until
+# the request to the metadata server fails. If the request succeeds and we get
+# a metadata change, it will kick off a node upgrade if any of the values
+# retrieved by a key in WATCH_KEYS changes. This process exits if a node upgrade
+# is kicked off. Note that for any case other than performing an upgrade, it
+# returns, and thus must be called repeatedly to continue checking for upgrades.
+#
+# Assumed vars:
+#   WATCH_KEYS
+#   CACHE_DIR
+#   CACHE_EXT
+#   LATEST_EXT
+function monitor-upgradability() {
+  # Wait for _any_ metadata to change.
+  log "Waiting for metadata changes."
+  curl -s "http://metadata/computeMetadata/v1/instance/attributes/?recursive=true&wait_for_change=true" -H "Metadata-Flavor: Google"
+  result=$?
+  if [[ ${result} -ne 0 ]]; then
+    log "Metadata request failed; returned ${result}"
+    return
+  fi
+
+  # Get latest metadata.
+  log "Metadata changed; getting latest versions."
+  get-metadata "${LATEST_EXT}"
+
+  # Diff cache vs latest.
+  log "Diffing cached versus latest metadata."
+  local do_upgrade=0
+  for k in "${WATCH_KEYS[@]}"; do
+    diff "${CACHE_DIR}/${k}.${CACHE_EXT}" "${CACHE_DIR}/${k}.${LATEST_EXT}" > /dev/null
+    if [[ $? -ne 0 ]]; then
+      log "Metadata with key '${k}' differ."
+      do_upgrade=1
+      break
+    fi
+  done
+
+  # Don't upgrade if no metadata changed.
+  if [[ ${do_upgrade} -eq 0 ]]; then
+    log "No metadata differences found; will not upgrade."
+    return
+  fi
+
+  # Kick off a node upgrade.
+  log "Relevant metadata differ; performing node upgrade."
+  upgrade-node
+}
+
+# upgrade-node launches a node upgrade, assuming that the script to do so is
+# found in ${CACHE_DIR}/startup-script.${LATEST_EXT}, then exits.
+#
+# Assumed vars:
+#   CACHE_DIR
+#   LATEST_EXT
+function upgrade-node() {
+  log "Running ${CACHE_DIR}/startup-script.${LATEST_EXT}"
+  # Wait 1 second (generous) for this upgrader to exit.
+  sudo /bin/bash -c "sleep 1; /bin/bash '${CACHE_DIR}/startup-script.${LATEST_EXT}' --push" &
```

Comment:
It's possible it's getting run by an arbitrary daemon user, but we don't create an explicit one for this service, so I'm at a loss for what it would be, and it's amazing that it has sudo privs if so.